File: pkg/model/api/v1/common/regexp.go
// Copyright 2021 The Perses Authors
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package common

import (
	"encoding/json"
	"regexp"
)

// Regexp encapsulates a regexp.Regexp and makes it JSON/YAML marshalable.
type Regexp struct {
	*regexp.Regexp
	original string
}

// NewRegexp creates a new anchored Regexp and returns an error if the
// passed-in regular expression does not compile.
func NewRegexp(s string) (Regexp, error) {
	regex, err := regexp.Compile(s)
	return Regexp{
		Regexp:   regex,
		original: s,
	}, err
}

// MustNewRegexp works like NewRegexp, but panics if the regular expression does not compile.
func MustNewRegexp(s string) Regexp {
	re, err := NewRegexp(s)
	if err != nil {
		panic(err)
	}
	return re
}

func (re *Regexp) UnmarshalJSON(data []byte) error {
	var s string
	if err := json.Unmarshal(data, &s); err != nil {
		return err
	}
	return re.validate(s)
}

func (re *Regexp) UnmarshalYAML(unmarshal func(interface{}) error) error {
	var s string
	if err := unmarshal(&s); err != nil {
		return err
	}
	return re.validate(s)
}

func (re Regexp) MarshalJSON() ([]byte, error) {
	if len(re.original) > 0 {
		return json.Marshal(re.original)
	}
	return nil, nil
}

func (re Regexp) MarshalYAML() (interface{}, error) {
	if len(re.original) > 0 {
		return re.original, nil
	}
	return nil, nil
}

func (re *Regexp) validate(s string) error {
	r, err := NewRegexp(s)
	if err != nil {
		return err
	}
	*re = r
	return nil
}

File: ui/core/src/utils/regexp.ts
// Copyright 2024 The Perses Authors
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/**
 * Checks if a string is a regex pattern.
 */
export function isRegexPattern(input: string | null | undefined): boolean {
  return Boolean(input?.startsWith('/'));
}

/**
 * Converts a string to RegExp. Handles both regular strings and regex patterns.
 * For regular strings, creates exact match regex.
 * For patterns like "/pattern/flags", creates corresponding RegExp.
 */
export function createRegexFromString(input: string): RegExp {
  if (!input) {
    throw new Error('Input string cannot be empty');
  }

  if (!isRegexPattern(input)) {
    return new RegExp(`^${input}$`);
  }

  const regexPattern = /^\/(.+)\/([gimy]*)$/;
  const matches = input.match(regexPattern);

  if (!matches) {
    throw new Error(`Invalid regular expression format: ${input}`);
  }

  const [, pattern = '', flags = ''] = matches;

  try {
    return new RegExp(pattern, flags);
  } catch (error) {
    throw new Error(`Failed to create RegExp ${error}`);
  }
}
